src/shims/unix/fs.rs

```diff
@@ -967,26 +965,21 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             builder.mode(mode);
         }
 
-        let result = builder.create(path).map(|_| 0i32);
-
-        Ok(Scalar::from_i32(this.try_unwrap_io_result(result)?))
+        Ok(builder.create(path))
     }
 
-    fn rmdir(&mut self, path_op: &OpTy<'tcx>) -> InterpResult<'tcx, Scalar> {
+    fn rmdir(&mut self, path_op: &OpTy<'tcx>) -> InterpResult<'tcx, io::Result<()>> {
         let this = self.eval_context_mut();
 
         let path = this.read_path_from_c_str(this.read_pointer(path_op)?)?;
 
         // Reject if isolation is enabled.
         if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
             this.reject_in_isolation("`rmdir`", reject_with)?;
-            this.set_last_error_from_io_error(ErrorKind::PermissionDenied.into())?;
-            return Ok(Scalar::from_i32(-1));
+            return Ok(Err(ErrorKind::PermissionDenied.into()));
```

Comment:
For instance, this could become `return Ok(this.return_io_error(ErrorKind::PermissionDenied.into()))` where `return_io_error` sets the errno and then returns `-1`.

Reply:
Yea, but I'd still need to say how big the scalar is supposed to be. Not the end of the world, but also not nice

Reply:
Ah right, the type...

Another option might be to pass `dest` to these functions and not have them return anything. A function that blocks usually already needs to do that anyway. Then they can use your new `write_io_result` internally for when they naturally have an `io::Result`, and some `write_io_error` for cases like this where we hard-code an error.